src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+    close(new_child->pipe);
+    log_println(6, "Too many clients, freeing child=0x%x", new_child);
+    free(new_child);
+    new_child = NULL;
+  } else {
+    // The server is not overloaded, so queue up the new client
+    if (*head == NULL) {
+      *head = *tail = new_child;
     } else {
-      // Nothing is in the queue, so wait forever until a new connection
-      // request arrives
-      log_println(3, "Timer not running, waiting for new connection");
-      mclients = 0;
-      do {
-        retcode = select(listenfd + 1, &rfd, NULL, NULL, NULL);
-        // retry if a signal caused the select() to exit
-      } while ((retcode == -1) && (errno == EINTR));
+      (*tail)->next = new_child;
+      *tail = new_child;
     }
+  }
+}
 
-    if (retcode < 0) {
-      // an interrupt or signal caused the select() to exit, go back and
-      // start over
-      log_println(5, "Select exited with return code = %d", retcode);
-      continue;
-    }
+/**
+ * Returns true or false, depending on whether the client's process is dead.
+ */
+int is_live_client(struct ndtchild *client) {
+  siginfo_t client_status;
+  int rv;
+  client_status.si_pid = 0;
+  rv = waitid(P_PID, client->pid, &client_status, WEXITED | WSTOPPED | WNOHANG);
+  return (rv == 0 && client_status.si_pid == 0);
+}
 
-    if (retcode == 0) {  // select exited due to timer expiration
-      log_println(3, "Timer expired while waiting for a new connection");
-      /* if ((waiting > 0) && (testing == 0)) */
-      if (multiple == 0) {
-        if ((waiting > 0) && (testing == 0)) goto ChldRdy;
-        continue;
-      } else {
-        if (waiting > 0) goto ChldRdy;
-        continue;
-      }
+/** Goes from head to tail, reaping dead clients from the queue of clients and
+ * then updating all clients regarding their (possibly new) queue position. */
+void queue_maintenance(struct ndtchild **head, struct ndtchild **tail) {
+  ParentChildSignal child_message;
+  int queue_position;
+  int client_count = 0;
+  int running_count = 0;
+  int max_simultaneous_tests;
+  struct ndtchild *tmp_ptr, *prev_ptr;
+  // Walk the list of clients, reaping dead clients.
+  tmp_ptr = *head;
+  prev_ptr = NULL;
+  while (tmp_ptr != NULL) {
+    if (is_live_client(tmp_ptr)) {
+      prev_ptr = tmp_ptr;
+      tmp_ptr = tmp_ptr->next;
     } else {
-      log_println(3,
-                  "New connection received, waiting for accept() to complete");
-      if ((waiting > 0) && (testing == 0)) {
-        // no clients waiting, no test in progress
-        goto ChldRdy;
-      }
-      clilen = sizeof(cli_addr);
-      memset(&cli_addr, 0, clilen);
-      log_println(6, "Select() found %d clients ready, highest fd=%d", retcode,
-                  listenfd);
-      if (retcode > 1) {
-        for (i = 3; i <= listenfd; i++) {
-          if (FD_ISSET(i, &rfd)) {
-            listenfd = i;
-            break;
-          }
-        }
+      // Remove tmp_ptr from the linked list
```

Comment:
Also doesn't update last_live_child->next when we reap the dead_child (all this dead_child language is pretty grim - whenever I talked about killing things, I've generally tried to refer to clients rather than children.  It caused a dialogue shift, but made the whole thing a lot less offputting)